### PR TITLE
Enable 12m and 24m metrics

### PR DESCRIPTION
**Pull request by Lukewh**
Fixes https://github.com/canonical-websites/snapcraft.io/issues/1359

Added "Past year" and "Past 2 years" options.

## QA
- Pull the branch
- `./run`
- Visit http://0.0.0.0:8004/<snap_name>/metrics
- Change the dropdown to Past year or Past 2 years

### CODE REVIEW

**Comment by Lukewh** [comment]
@evandandrea that's why this is a [WIP] PR :wink:

**Comment by evandandrea** [comment]
Whoops!

**Comment by matthewpaulthomas** [comment]
What date is the earliest data for any snap? Would be nice if any “2 years” option appears only once it becomes useful.

**Comment by Lukewh** [comment]
Blocked on https://bugs.launchpad.net/snapstore/+bug/1808521

File: static/js/publisher/metrics/graphs/activeDevices.js [hunk 1]
```diff
@@ -210,7 +210,16 @@ function drawGraph(holderSelector, holder, activeDevices) {
 
   // Add the x axix
   let tickValues = [];
-  if (isMobile() && _data.length > 90) {
+  let tickFormat = "%b %e";
+
+  if (_data.length > 360) {
```
Comment (bartaz) [inline static/js/publisher/metrics/graphs/activeDevices.js]: Could this part get a bit of comment explaining what is being changed in these conditions?
Seems like formatting changes based on size of data we have? Why is something different happening on mobile?

**Comment by bartaz** [comment]
Works well, but for some big snaps (like core) it always returns gateway timeout for 2y period.

File: static/js/publisher/metrics/graphs/activeDevices.js (hunk 1 above)
Comment (Lukewh) [reply static/js/publisher/metrics/graphs/activeDevices.js]: :+1: I also want to add a comment to the python bit too

**Comment by Lukewh** [comment]
> Works well, but for some big snaps (like core) it always returns gateway timeout for 2y period.

@tbille is there anything we can do about the timeouts? People are still experiencing them quite regularly.

File: webapp/metrics/helper.py [hunk 2]
```diff
@@ -37,12 +37,15 @@ def build_metrics_json(
...
     if metric_bucket == "d":
-        start = end - relativedelta.relativedelta(days=metric_period)
+        start = end + relativedelta.relativedelta(days=-metric_period)
     elif metric_bucket == "m":
-        start = end - relativedelta.relativedelta(months=metric_period)
+        start = end + relativedelta.relativedelta(months=-metric_period)
+    elif metric_bucket == "y":
+        # Go back an extra day to ensure the granularity increases
+        start = end + relativedelta.relativedelta(
+            years=-metric_period, days=-1
```
Comment (tbille) [inline webapp/metrics/helper.py]: no -1 or : `{'error_list': [{'message': 'Filter interval larger than 365 days at index 0', 'code': 'invalid-date-interval', 'extra': {'index': 0, 'end': '2019-01-28', 'start': '2018-01-27', 'max_days': 365}}]}`

Comment (Lukewh) [reply webapp/metrics/helper.py]: The extra day ensures the granularity decreases. Either way, removing the -1 day will still cause an error for a 2 year window. I'll discuss with @cprov why this limit is in place and where it came from, as this worked last week.

**Pull request merged**